File: 684-redundant-connection/684-redundant-connection.java
class Solution {
    public int[] findRedundantConnection(int[][] edges) {
        
        UnionFind uf = new UnionFind(edges.length);
        
        for(int[] edge : edges){
            int node1 = edge[0], node2 = edge[1];
            if(uf.union(node1,node2) == 0) return edge;
        }
        
        return new int[]{0,0};
    }
}

class UnionFind{
    int[] root;
    int[] rank;
        
    public UnionFind(int size){
        root = new int[size + 1];
        rank = new int[size + 1];
        for(int i = 0; i <= size; i++) root[i] = i;
    }
    
    public int find(int elem){
        
        if(root[elem] == elem) return elem;
        
        return root[elem] = find(root[elem]); // with path compression 
    }
    
    public int union(int elem1, int elem2){
        
        int root1 = find(elem1);
        int root2 = find(elem2);
        
        if(root1 == root2) return 0;
        
        // union by rank
        if(rank[root1] > rank[root2]){
            root[root2] = root1;
        }
        else if(rank[root1] < rank[root2]){
            root[root1] = root2;
        }
        else{
            root[root2] = root1;
            rank[root1]++;
        }
        
        return 1;
        
        
    }
    
    
}

File: 684-redundant-connection/684-redundant-connection.py
class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        n = len(edges)
        uf = UnionFind(n + 1) 
        
        for u,v in edges:
            if not uf.union(u,v):
                return u,v
        
        
        

class UnionFind:
    
    def __init__(self,n):
        self.parents = [i for i in range(n)]
        self.ranks = [0] * n  
    
    def find(self,x):
        if self.parents[x] != x:
            self.parents[x] = self.find(self.parents[x]) # path compression 
        return self.parents[x] 
    
    def union(self,u,v):
        u_p = self.find(u) 
        v_p = self.find(v) 
        
        if u_p == v_p:
            return False 
        elif self.ranks[u_p] > self.ranks[v_p]:
            self.parents[v_p] = u_p 
        elif self.ranks[u_p] < self.ranks[v_p]:
            self.parents[u_p] = v_p 
        else:
            self.parents[v_p] = u_p 
            self.ranks[u_p] += 1 
        return True 